src/electrai/dataloader/qm9.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import annotations
+
+from pathlib import Path
+
+import numpy as np
+import torch
+from sklearn.model_selection import train_test_split
+from torch.utils.data import Dataset
+
+from .registry import register_data
+
+dtype_map = {"f32": torch.float32, "f16": torch.float16, "bf16": torch.bfloat16}
+
+
+class RhoRead:
+    def __init__(
+        self,
+        data_path: Path,
+        label_path: Path,
+        train_fraction: float,
+        random_state: int = 42,
+    ):
+        """
+        Parameters
+        ----------
+        data_path: path of input chgcar or elfcar files.
+        label_path: path of label chgcar or elfcar files.
+        train_fraction: fraction of the data used for training (0 to 1).
+        """
+        self.data_path = Path(data_path)
+        self.label_path = Path(label_path)
+        self.tf = train_fraction
+        self.rs = random_state
+
+    def data_split(self):
+        data_list = []
+
+        for mol_id in range(1, 200):
+            data = (
+                self.data_path / f"dsgdb9nsd_{mol_id:06d}" / "rho_22.npy",
+                self.label_path / f"dsgdb9nsd_{mol_id:06d}" / "rho_22.npy",
+                self.data_path / f"dsgdb9nsd_{mol_id:06d}" / "grid_sizes_22.dat",
+                self.label_path / f"dsgdb9nsd_{mol_id:06d}" / "grid_sizes_22.dat",
+            )
+            data_list.append(data)
+        train_data, test_data = train_test_split(
+            data_list, train_size=self.tf, random_state=self.rs
+        )
+        return train_data, test_data
+
+
+class RhoData(Dataset):
+    def __init__(
+        self,
+        data: list[tuple[Path, Path, Path, Path]],
+        data_precision: str,
+        data_augmentation=True,
+        downsample_data=1,
+        downsample_label=1,
+    ):
+        """
+        Parameters
+        ----------
+        data: list of (input voxel data, label voxel data, input gridsize, label gridsize) of length batch_size.
+        """
+        self.ds_data = downsample_data
+        self.ds_label = downsample_label
+        self.da = data_augmentation
+        self.data = data
+        self.data_precision = data_precision
+
+    def __len__(self):
+        return len(self.data)
+
+    def rotate_x(self, data_in):
```

Comment:
This doesn't need to happen in this PR but note for the future that there's a lot of duplicated code between dataloaders which should be abstracted.